### DISCUSSION TITLE
Cannot obtain asset key from RunFailureSensor

### DISCUSSION THREAD

**Discussion by cody-scott** [dagster-io/dagster#20404] [Q&A]
Hi. I am trying to write a test to target a specific set of asset keys to trigger a failure sensor on, but in my testing i'm not returning a valid asset key.

Here is the snippet i'm testing on.

```python
from dagster import (
    asset,
    DagsterInstance,
    build_run_status_sensor_context,
    Definitions,
    define_asset_job,
)

from dagster import RunFailureSensorContext

from dagster import run_failure_sensor


@run_failure_sensor
def this_will_fail_sensor(context: RunFailureSensorContext):
    [_.asset_key for _ in context.get_step_failure_events()]  # -> [None]


def test_triggers_on_hxm_fail():
    @asset(key_prefix=["abc", "failed"])
    def will_pass(): ...

    @asset(key_prefix=["abc", "failed"], deps=[will_pass])
    def will_fail():
        raise Exception("oh no")

    _job = define_asset_job("fail_job", [will_fail])

    instance = DagsterInstance.ephemeral()
    defs = Definitions(assets=[will_fail], jobs=[_job])
    result = defs.get_job_def("fail_job").execute_in_process(
        instance=instance, raise_on_error=False
    )

    dagster_run = result.dagster_run
    dagster_event = result.get_job_failure_event()

    context = build_run_status_sensor_context(
        sensor_name="this_will_fail_sensor",
        dagster_instance=instance,
        dagster_run=dagster_run,
        dagster_event=dagster_event,
    )
    this_will_fail_sensor(context)
```

My expectation is that i would be able to obtain the asset key for the failure, but it doesn't seem to be the case, they just return `None`.

My alternative though is that i would be able to use an `AsseyKey` and build up the target myself (in this case `abc__failed__will_fail`). Am I missing the proper approach here? 

I have looked at https://docs.dagster.io/concepts/partitions-schedules-sensors/sensors#testing-run-status-sensors and https://github.com/dagster-io/dagster/discussions/20256 but no luck here either,

**Comment by dpeng817** [comment]
Hey there - this is an unfortunate result of the original API being constructed before assets were the main abstraction in dagster. Step events do not directly map onto assets, since you can have a multi asset which executes in a single step. Instead, I think what you want is to retrieve the set of planned materializations from the run. Something like:
```python
@run_failure_sensor
def my_sensor(context):
    run_id = context.dagster_run.run_id
    events = context.instance.fetch_planned_materializatons(run_id)
```

Unfortunately, it's currently not possible without digging into internal APIs, and even then, there's no way to query on a per-run basis, since this is currently not performant for assets.

I'll open an issue about this in a more general format of retrieving the failed assets from a run.
